### PR TITLE
3070 filtering applicants by classification should use generics on backend

### LINKED ISSUE #3070: Filtering applicants by classification should use generics on backend
On the Pool Candidate and Applicant search, filtering by Classification currently has a implementation on the backend where it matches against expectedClassifications **or** against expected Salary.

Leave the Pool Candidate search alone, but update the User/Applicant search so that classifications match classification to
- expectedClassifications OR
- expectedSalary OR
- expectedGenericJobTitle

### PR DESCRIPTION
**Pull request by RVany**
Resolve #3070 
On the Pool Candidate and Applicant search, filtering by Classification currently has a implementation on the backend where it matches against expectedClassifications **or** against expected Salary.

Leave the Pool Candidate search alone, but update the User/Applicant search so that classifications match classification to
- expectedClassifications OR
- expectedSalary OR
- expectedGenericJobTitle

### CODE REVIEW

File: api/app/Models/User.php [hunk 1]
```diff
@@ -348,6 +348,40 @@ public function filterByClassifications(Builder $query, array $classifications):
...
 
             $this->orFilterByClassificationToSalary($query, $classifications);
+           $this->orfilterByClassificationToGenericJobTitles($query, $classifications);
+
+
+        });
+
+        return $query;
+    }
+     public function orfilterByClassificationToGenericJobTitles(Builder $query, array $classifications): Builder
```
Comment (tristan-orourke) [inline api/app/Models/User.php:358]: Follow casing convention for method names
```suggestion
           $this->orFilterByClassificationToGenericJobTitles($query, $classifications);


        });

        return $query;
    }
     public function orFilterByClassificationToGenericJobTitles(Builder $query, array $classifications): Builder
```

File: api/app/Models/User.php [hunk 2]
```diff
@@ -348,6 +348,40 @@ public function filterByClassifications(Builder $query, array $classifications):
             });
 
             $this->orFilterByClassificationToSalary($query, $classifications);
+           $this->orfilterByClassificationToGenericJobTitles($query, $classifications);
```
Comment (tristan-orourke) [inline api/app/Models/User.php:351]: Indentation inconsistency
```suggestion
            $this->orfilterByClassificationToGenericJobTitles($query, $classifications);
```

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
Besides some formatting improvements, the code looks right to me! We should just add some tests to UserTest.php before merging this. If you don't have experience with phpunit, I can help with that.

**Comment by RVany** [comment]
> Besides some formatting improvements, the code looks right to me! We should just add some tests to UserTest.php before merging this. If you don't have experience with phpunit, I can help with that.

Ok. Thank you!

**Comment by vd1992** [comment]
image

**Comment by petertgiles** [comment]
I'm getting this error in my Laravel log from the search requests:
```
[2022-06-27 20:56:57] local.ERROR: Method 'filterByClassifications' does not exist on class 'App\Models\User' {"userId":"d36f699b-cbf9-4104-bec0-7d724e27be61","exception":"[object] (Nuwave\\Lighthouse\\Exceptions\\DefinitionException(code: 0): Method 'filterByClassifications' does not exist on class 'App\\Models\\User' at /var/www/html/api/vendor/nuwave/lighthouse/src/Support/Utils.php:53)
```
Are you seeing this?

**Comment by vd1992** [comment]
hm, it runs now but I'm trying to wrap my head around the query so I don't think I can really test if the results are what they should be
@petertgiles @JamesHuf can I request either of you to put the query through its paces?

**Comment by JamesHuf** [comment]
> hm, it runs now but I'm trying to wrap my head around the query so I don't think I can really test if the results are what they should be @petertgiles @JamesHuf can I request either of you to put the query through its paces?

I'm looking at it now. 😃

**Review by JamesHuf — APPROVED** [review]
Everything seems to work as expected. 👍

**Pull request merged**